onnxoptimizer/passes/eliminate_nop_gather_shape.h

```diff
@@ -0,0 +1,101 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// ATTENTION: The code in this file is highly EXPERIMENTAL.
+// Adventurous users should note that the APIs will probably change.
+
+#pragma once
+
+#include "onnx/defs/tensor_util.h"
+#include "onnxoptimizer/pass.h"
+
+namespace ONNX_NAMESPACE {
+namespace optimization {
+
+struct EliminateNopGatherShape final : public PredicateBasedPass {
+  explicit EliminateNopGatherShape()
+      : PredicateBasedPass(PassType::Nop, PassEfficiency::Complete,
+                           PassOptimizationType::Compute) {}
+
+  std::string getPassName() const override {
+    return "eliminate_nop_gather_shape";
+  }
+
+  bool patternMatchPredicate(Node *node) override {
+    return node->kind() == Symbol("Gather") &&
+           node->inputs()[0]->node()->kind() == Symbol("Shape") &&
+           node->inputs()[0]->node()->input()->has_sizes();
+  }
+
+  bool runTransform(Node *node, Graph &graph,
+                    NodeDestroyType &destroy_current) override {
+    auto *x = node->inputs()[0];
+    auto *indices = node->inputs()[1];
+    const auto &initializer_names = graph.initializer_names();
+    Node *shape = x->node();
+    const auto &dims = shape->input()->sizes();
+
+    const Tensor *indices_tensor = nullptr;
+    if (indices->node()->kind() == kConstant) {
+      indices_tensor = &indices->node()->t(kvalue);
+    } else if (indices->node()->kind() == kParam &&
+               std::find(initializer_names.cbegin(), initializer_names.cend(),
+                         indices->uniqueName()) != initializer_names.cend()) {
+      indices_tensor = &*graph.getInitializer(indices->uniqueName());
+    } else {
+      return false;
+    }
+
+    int indices_val;
+    if (indices_tensor->elem_type() ==
+        ONNX_NAMESPACE::TensorProto_DataType_INT32) {
+      indices_val = ParseData<int32_t>(indices_tensor)[0];
```

Comment:
From documents Gather indices can have any rank instead of either 0 or 1. Maybe we should check the rank and return false for indices whose rank greater than 1?

Reply:
the rank of shape's output is 1D